etsin_finder/frontend/js/stores/view/qvain/qvain.submit.js

```diff
@@ -0,0 +1,282 @@
+    this.isLoading = state
+  }
+
+  @action setError = error => {
+    this.error = error
+  }
+
+  @action clearResponse = () => {
+    this.response = null
+  }
+
+  @action closeUseDoiModal = () => {
+    this.useDoiModalIsOpen = false
+  }
+
+  @action openUseDoiModal = () => {
+    this.useDoiModalIsOpen = true
+  }
+
+  @computed get submitType() {
+    const { original } = this.Qvain
+    if (!original) {
+      return DATASET_STATE.NEW
+    }
+    if (original.state === 'draft') {
+      if (original.draft_of) return DATASET_STATE.UNPUBLISHED_DRAFT
+
+      return DATASET_STATE.DRAFT
+    }
+
+    return DATASET_STATE.PUBLISHED
+  }
+
+  @action submitDraft = async () => {
+    switch (this.submitType) {
+      case DATASET_STATE.NEW:
+        await this.exec(this.createNewDraft)
+        return
+      case DATASET_STATE.DRAFT:
+      case DATASET_STATE.UNPUBLISHED_DRAFT:
+        await this.exec(this.updateDataset)
+        return
+      case DATASET_STATE.PUBLISHED:
+        await this.exec(this.savePublishedAsDraft)
+        return
+      default:
+        console.error('Unknown submit status')
+        throw new Error('Unknown submit status')
+    }
+  }
+
+  @action submitPublish = async () => {
+    switch (this.submitType) {
+      case DATASET_STATE.NEW:
+        await this.exec(this.publishNewDataset)
+        return
+      case DATASET_STATE.DRAFT:
+        await this.exec(this.publishDraft)
+        return
+      case DATASET_STATE.UNPUBLISHED_DRAFT:
+        await this.exec(this.mergeDraft)
+        return
+      case DATASET_STATE.PUBLISHED:
+        await this.exec(this.republish)
+        return
+      default:
+        console.error('Unknown submit status')
+        throw new Error('Unknown submit status')
+    }
+  }
+
+  @action exec = async (submitFunction, schema = qvainFormSchema) => {
+    const {
+      OtherIdentifiers: { cleanupBeforeBackend },
+      editDataset,
+      setChanged,
+    } = this.Qvain
+
+    if (!cleanupBeforeBackend()) return
+    if (!(await this.promptProvenancesAndCleanOtherIdentifiers())) return
+
+    this.closeUseDoiModal()
+    const dataset = this.prepareDataset()
+    const { fileActions, metadataActions, newCumulativeState } = this.prepareActions()
+
+    try {
+      await schema.validate(dataset)
+      // kind of overrules the other validation errors.
+      // Will be fixed in CSCFAIRMETA-542.
+      await this.checkDoiCompability(dataset)
+      this.setLoading(true)
+      const { data } = await submitFunction(dataset)
+      await this.updateFiles(data.identifier, fileActions, metadataActions)
+      setChanged(false)
+
+      if (fileActions || metadataActions || newCumulativeState) {
+        // Files changed, get updated dataset
+        const url = urls.v2.dataset(data.identifier)
+        const updatedResponse = await axios.get(url)
+        await editDataset(updatedResponse.data)
+      } else {
+        await editDataset(data)
+      }
+      this.response = data
```

Comment:
Because code after `await` is no longer in an action, this needs `runInAction(() => { this.response = data })` or maybe preferably you could create a helper `@action setResponse(response)`.